### DISCUSSION TITLE
Negative latency audio timing on Windows

### DISCUSSION THREAD

**Discussion by SonoSooS** [ppy/osu#18328] [Q&A]
I always try to synchronize the music with the sound of my keyboard, but somehow that means that I always press early, so I decided to measure it in-game.

The test was conduced by the very scientific method of pressing the keyboard key really hard and really fast, so it makes a really loud sound when it activates, and I tried my best to sync the key clack with the metronome tick. I used No Fail, Target, Difficulty Adjust (CS=0), and I muted hit sounds and the music, so they wouldn't influence the test. I did also account for key travel time, by making it activate on bottom-out. Despite hitting slightly late in the test, trying to not hit too early, it still averaged out on a negative latency.

How is this possible?

![image](https://user-images.githubusercontent.com/7823099/169175113-76cc7086-1148-4658-955b-b44ad33e1e99.png)
![image](https://user-images.githubusercontent.com/7823099/169176514-85011b6e-7dea-4d2a-8cb6-a05f3a00b432.png)

**Comment by peppy** [comment]
adjust your universal offset to account for this? i'm not sure what you're asking here.

**Reply by SonoSooS** [reply]
I'm just saying that this should be normally impossible. I have universal offset at 0ms, but I can't think of a way where I could hear the metronome sound into the future.

**Comment by bdach** [comment]
what OS is this? there is a platform-specific adjustment on windows that may be causing this.

**Reply by peppy** [reply]
and beatmap timing and what not

**Reply by SonoSooS** [reply]
This only seems to happen on Windows, yes.

Does beatmap timing really affect this? I thought practice mode with everything but the metronome muted would be a good test subject.

**Reply by bdach** [reply]
you're probably seeing the effects of this, then

https://github.com/ppy/osu/blob/77b4bd97f8a73f0c404a56ff408250af3cd671b0/osu.Game/Screens/Play/MasterGameplayClockContainer.cs#L195-L197

**Reply by SonoSooS** [reply]
I set my universal offset to 15ms, and now I'm now hitting the notes dead-on on difficult technically-timed maps!

Good to be confirmed that it was not my fault all this time, it was driving me slightly insane that I suck on difficult maps for seemingly no reason.

Edit: if I set that platform offset to 0, then I must set my offset to somewhere between +26ms and +30ms.

**Comment by jansenmtan** [comment]
osu! ranked beatmaps are, on average, timed to be 26 ms off-time: 26 ms late in terms of judgement, 26 ms early in terms of audio.
I base this conclusion off the data I'm collecting in [this spreadsheet](https://docs.google.com/spreadsheets/d/1eI3QlhFUs2ivXgZGvil2WZr3MAnHuzLv9_nRz2piqoM/edit?usp=sharing) I made by measuring the offset error in the lazer editor (which currently does not incorporate user-set offsets).

This is why audio offset must be set to at least -26 ms on non-Windows platforms. On Windows, it must be set another -15 ms early due to the offset fudging bdach mentions.

**Reply by frenzibyte** [reply]
Can you specify your methodology in finding this offset error? And keep in mind that lazer editor doesn't have the windows offset applied currently.

**Reply by peppy** [reply]
I've tested a few beatmaps from your spreadsheet and I can attest that this is not the case for me on macOS and linux. This is likely your local audio pipeline having latency and a prime example of needing to adjust your universal offset setting.

**Reply by jansenmtan** [reply]
The testing methodology is shown in the following video (done on Windows):

https://user-images.githubusercontent.com/36340429/180233860-143ab190-ebb7-436d-99c1-c029eb64ad86.mp4


In the video, I set the timing offset of the beatmap to the snare, which happens to be 26 ms later, relative to the original timing offset. Searchlight on my spreadsheet has a beatmap timing offset error [value of -26](https://docs.google.com/spreadsheets/d/1eI3QlhFUs2ivXgZGvil2WZr3MAnHuzLv9_nRz2piqoM/edit#gid=1705391314&range=A22).
Some genres of music are easier to find the beatmap timing offset error than others. eg, values found from electronic music can be trusted more than metal or rock.

Another video, done on Linux this time. This phenomenon is platform independent.

https://user-images.githubusercontent.com/36340429/180244496-5411755b-7327-4132-9a44-f98bd8d7cd2c.mp4

Also, my UO on Windows is -42 and my UO on Linux is -28. Yet, the BTOE is the same on both platforms.

**Reply by peppy** [reply]
Can you confirm your issue is fixed by adjusting your Universal Offset? If so, I think that is your best bet for now.

**Reply by jansenmtan** [reply]
> I've tested a few beatmaps from your spreadsheet and I can attest that this is not the case for me on macOS and linux. This is likely your local audio pipeline having latency and a prime example of needing to adjust your universal offset setting.

The "beatmap timing offset error" values in the spreadsheet are independent of gameplay and universal/global/audio offset because they are found in the editor.
Even if #11400 is implemented, people who set the audio offset to the true audio latency of their machines will find a massive timing offset error in the editor, with an average value of:
$$o = \verb|average_ranked_timing_error| + \verb|audio_latency| = -26 + \verb|audio_latency|$$ To fix this with lazer as it is would involve either having to set the "local offset" to $o$ ms early, or having to set the audio offset to a value of $-26 - \verb|audio_latency|$. 
If you change the local offset, then it's just annoying cause most (likely >99% of) ranked beatmaps have a BTOE greater than 10 in magnitude. 
If you change the audio offset, then this makes the term "audio offset" effectively meaningless and misleading, and requires a different term, or a new offset altogether.

**Reply by peppy** [reply]
Your sample size of one is an issue here. I cannot reproduce the same 26ms offset across three completely different machines on three different OSes.

The offset will be applied to the editor soon. You can track this at https://github.com/ppy/osu/issues/11400. Please wait for this.

**Reply by jansenmtan** [reply]
My observations are made in conditions similar to that of SonoSooS, ie, without hitsounding. If anyone is to reproduce these results + come to the same conclusions, it should be with hitsounding off. Also, they should not be using closed-back headphones or IEMs (they should hear their taps), and use a mechanical keyboard.

---

This conclusion is based off of around 100 hours of gameplay and study. I wrote about one beatmap, but I observed and measured offsets of hundreds of them. Even the spreadsheet only contains just the fraction of beatmap offsets I bothered to write down.
I didn't even know about the code at https://github.com/ppy/osu/blob/77b4bd97f8a73f0c404a56ff408250af3cd671b0/osu.Game/Screens/Play/MasterGameplayClockContainer.cs#L195-L197 until July 17, yet I observed and measured it to the millisecond on June 19. I wouldn't be saying any of this if it wasn't scientifically derived. I trust my conclusions.

---

> The offset will be applied to the editor soon. You can track this at https://github.com/ppy/osu/issues/11400. Please wait for this.

If I understand correctly, #11400 would not lead me to changing my audio offset setting. It would make the waveform in the editor "line up" correctly with the beatmap, given my current offset settings. I would still need my current offset settings at their values.

**Reply by peppy** [reply]
I'll keep this simple:

The current settings we have ensure that hitsounds line up with gameplay
- if they don't for you, then please adjust your universal offset
- if you don't use hitsounds, that's fine. use the same universal offset to align gameplay

We will continue to assess audio offsets and expose them in the most intuitive way. The current goal is to have 0ms offset be correctly playable for osu!'s extensive catalog of beatmaps, and stay compatible with countless official and non-official osu! clients.

Thanks for your understanding!
